src/dune/file_tree.ml

```diff
@@ -306,76 +370,123 @@ module rec Memoized : sig
-                     | None -> Some path
-                     | Some first_path ->
-                       User_error.raise
-                         [ Pp.textf
-                             "Path %s has already been scanned. Cannot scan it \
-                              again through symlink %s"
-                             (Path.Source.to_string_maybe_quoted first_path)
-                             (Path.Source.to_string_maybe_quoted path)
-                         ])
-                 in
-                 String.Map.add_exn dirs_visited_acc fn new_dirs_visited
-             in
-             let sub_dir =
-               let sub_dir_as_t = find_dir_raw path in
-               { Dir0.sub_dir_status = dir_status; sub_dir_as_t }
-             in
-             let subdirs = String.Map.set subdirs fn sub_dir in
-             (dirs_visited_acc, subdirs))
+  module Get_subdir : sig
+    (** Get all the sub directories of [path].*)
+    val all :
+         dirs_visited:Dirs_visited.t
+      -> dirs:(string * Path.Source.t * File.t) list
+      -> sub_dirs:Predicate_lang.Glob.t Sub_dirs.Status.Map.t
+      -> parent_status:Sub_dirs.Status.t
+      -> dune_file:Dune_file.t option (** to interpret [(chdir ..)] stanzas *)
+      -> path:Path.Source.t
+      -> Dirs_visited.Per_fn.t * Dir0.sub_dir String.Map.t
+  end = struct
+    let status ~status_map ~(parent_status : Sub_dirs.Status.t) dir :
+        Sub_dirs.Status.t option =
+      let status = Sub_dirs.status status_map ~dir in
+      match status with
+      | Ignored -> None
+      | Status status ->
+        Some
+          ( match (parent_status, status) with
+          | Data_only, _ -> Data_only
+          | Vendored, Normal -> Vendored
+          | _, _ -> status )
+
+    let make_subdir ~dir_status ~virtual_ path =
+      let sub_dir_as_t = find_dir_raw path in
+      { Dir0.sub_dir_status = dir_status; sub_dir_as_t; virtual_ }
+
+    let physical ~dirs_visited ~dirs ~sub_dirs ~parent_status =
+      let status_map =
+        Sub_dirs.eval sub_dirs ~dirs:(List.map ~f:(fun (a, _, _) -> a) dirs)
+      in
+      List.fold_left dirs ~init:(Dirs_visited.Per_fn.init, String.Map.empty)
+        ~f:(fun (dirs_visited_acc, subdirs) ((fn, path, _) as dir) ->
+          match status ~status_map ~parent_status fn with
+          | None -> (dirs_visited_acc, subdirs)
+          | Some dir_status ->
+            let dirs_visited_acc =
+              Dirs_visited.Per_fn.add dirs_visited_acc dirs_visited dir
+            in
+            let sub_dir = make_subdir ~dir_status ~virtual_:false path in
+            let subdirs = String.Map.set subdirs fn sub_dir in
+            (dirs_visited_acc, subdirs))
+
+    (* virtual sub directories are those that aren't on the file system but are
+       mentioned in the dune file via [(chdir ..)]. We must handle their status
+       explicitly, but they do contribute to the visited directories map as they
```

Comment:
FWIW, I don't understand the comment about cycles. (what cycles is this about?)

Reply:
I read it again and yeah this comment doesn't make much sense. I got rid of it.

My original goal was to explain why `~dirs_visited` isn't passed like it is in `phsyical`. In retrospect, it's obvious and the comment just confuses matters.